File: shared/src/types/task.ts
export interface Task {
  id: string;
  projectId: string;
  type: TaskType;
  status: TaskStatus;
  payload: TaskPayload;
  result?: TaskResult;
  createdAt: Date;
  updatedAt: Date;
  startedAt?: Date;
  completedAt?: Date;
}

export enum TaskType {
  CODE_REVIEW = 'CODE_REVIEW',
  ISSUE_ANALYSIS = 'ISSUE_ANALYSIS',
  PR_CREATION = 'PR_CREATION',
  CODE_GENERATION = 'CODE_GENERATION',
  DOCUMENTATION = 'DOCUMENTATION',
}

export enum TaskStatus {
  PENDING = 'PENDING',
  QUEUED = 'QUEUED',
  PROCESSING = 'PROCESSING',
  COMPLETED = 'COMPLETED',
  FAILED = 'FAILED',
  CANCELLED = 'CANCELLED',
}

export interface TaskPayload {
  [key: string]: any;
}

export interface TaskResult {
  success: boolean;
  data?: any;
  error?: string;
  artifacts?: string[];
}

File: backend/app/models/task.py
import enum
from datetime import datetime

from sqlalchemy import JSON, DateTime, Enum, ForeignKey, String
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.core.database import Base


class TaskType(str, enum.Enum):
    CODE_REVIEW = "CODE_REVIEW"
    ISSUE_ANALYSIS = "ISSUE_ANALYSIS"
    PR_CREATION = "PR_CREATION"
    CODE_GENERATION = "CODE_GENERATION"
    DOCUMENTATION = "DOCUMENTATION"


class TaskStatus(str, enum.Enum):
    PENDING = "PENDING"
    QUEUED = "QUEUED"
    PROCESSING = "PROCESSING"
    COMPLETED = "COMPLETED"
    FAILED = "FAILED"
    CANCELLED = "CANCELLED"


class Task(Base):
    __tablename__ = "tasks"

    id: Mapped[str] = mapped_column(String, primary_key=True)
    project_id: Mapped[str] = mapped_column(ForeignKey("projects.id"))
    type: Mapped[TaskType] = mapped_column(Enum(TaskType))
    status: Mapped[TaskStatus] = mapped_column(Enum(TaskStatus), default=TaskStatus.PENDING)
    payload: Mapped[dict] = mapped_column(JSON)
    result: Mapped[dict | None] = mapped_column(JSON, nullable=True)
    created_at: Mapped[datetime] = mapped_column(DateTime, default=datetime.utcnow)
    updated_at: Mapped[datetime] = mapped_column(
        DateTime, default=datetime.utcnow, onupdate=datetime.utcnow
    )
    started_at: Mapped[datetime | None] = mapped_column(DateTime, nullable=True)
    completed_at: Mapped[datetime | None] = mapped_column(DateTime, nullable=True)

    # Relationships
    project: Mapped["Project"] = relationship(back_populates="tasks")
